### PR TITLE
feat: Get org by `id` or `login`

### LINKED ISSUE #1722: [Story] Enable to get single org from the contributions objects returned by the Challenge service
### What projects is this story for?

OpenChallenges

### As a user, I want

As a UI developer, I want to be able to fetch org objects from the contributions objects that the challenge service returns so that I can display org cards on the challenge profile page.

### Description

There is an inconsistency on the identifier used to manage organization objects.

- Single organization are fetch by their login
- The contribution object returned by the challenge service includes the organization ID

This inconsistency makes it difficult to show a selection of organization cards, e.g. the orgs that contributed to the organization of a challenge on a challenge profile page.

Potential solutions are:

- Update the organization service to fetch orgs by their ID instead of their login OR
- Update the contribution object to include the org logins instead of their IDs

Considerations:

- If we want to keep the URL of org profile pages user- and SE-friendly like `openchallenges.io/{org_login}`, then we need a way to fetch an organization by their login.
- Using org ID instead of login is more performant: logins are typically longer than ID (integer) and may potentially change, e.g. when an organization changes its name (e.g. rebranding or an org is bought by another org).

### Acceptance criteria

_No response_

### Tasks

_No response_

### Anything else?

_No response_

### Have you linked this story to a GitHub Project?

- [X] I have linked this story to a GitHub Project and set its metadata.

### PR DESCRIPTION
**Pull request by tschaffter**
Closes #1722

## Changelog

- Update endpoint to access `id` value in addition to `login` values
- Requires org logins to start with a-z
- Update the OC API client for Angular

## Notes

- The motivation for requiring org logins to start with a-z is so that there is never an org login that may be an org ID too, which would otherwise lead to ambiguity when retrieving the organization.

### Get an organization by login

```
GET {{basePath}}/organizations/dream

HTTP/1.1 200
{
  "id": 1,
  "name": "DREAM",
  "email": "dream@sagebionetworks.org",
  "login": "dream",
  "description": "Together, we share a vision to enable individuals and groups to collaborate openly so that the “wisdom of the crowd” provides the greatest impact on science and human health.",
  "avatarKey": "logo/dream.png",
  "websiteUrl": "https://dreamchallenges.org/",
  "challengeCount": 47,
  "createdAt": "2023-07-20T21:46:02Z",
  "updatedAt": "2023-07-20T21:46:02Z",
  "acronym": "DREAM"
}
```

### Get an organization by ID

```
GET {{basePath}}/organizations/1

HTTP/1.1 200 
{
  "id": 1,
  "name": "DREAM",
  "email": "dream@sagebionetworks.org",
  "login": "dream",
  "description": "Together, we share a vision to enable individuals and groups to collaborate openly so that the “wisdom of the crowd” provides the greatest impact on science and human health.",
  "avatarKey": "logo/dream.png",
  "websiteUrl": "https://dreamchallenges.org/",
  "challengeCount": 47,
  "createdAt": "2023-07-20T21:46:02Z",
  "updatedAt": "2023-07-20T21:46:02Z",
  "acronym": "DREAM"
}
```

### Requiring org login to start with `a-Z`

The organization service fails to load if we replace the login of DREAM by `4dream`:

```console
$ nx serve openchallenges-organization-service
...
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.handleException(DefaultSqlScriptExecutor.java:275)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.executeStatement(DefaultSqlScriptExecutor.java:222)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.execute(DefaultSqlScriptExecutor.java:126)
        at org.flywaydb.core.internal.resolver.sql.SqlMigrationExecutor.executeOnce(SqlMigrationExecutor.java:68)
        at org.flywaydb.core.internal.resolver.sql.SqlMigrationExecutor.lambda$execute$0(SqlMigrationExecutor.java:57)
        at org.flywaydb.core.internal.database.DefaultExecutionStrategy.execute(DefaultExecutionStrategy.java:27)
        at org.flywaydb.core.internal.resolver.sql.SqlMigrationExecutor.execute(SqlMigrationExecutor.java:56)
        at org.flywaydb.core.internal.command.DbMigrate.doMigrateGroup(DbMigrate.java:376)
        ... 153 common frames omitted
Caused by: java.sql.SQLIntegrityConstraintViolationException: CONSTRAINT `login_check` failed for `organization_service`.`organization`
        at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:117)
        at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:122)
        at com.mysql.cj.jdbc.StatementImpl.executeInternal(StatementImpl.java:763)
        at com.mysql.cj.jdbc.StatementImpl.execute(StatementImpl.java:648)
        at com.zaxxer.hikari.pool.ProxyStatement.execute(ProxyStatement.java:94)
        at com.zaxxer.hikari.pool.HikariProxyStatement.execute(HikariProxyStatement.java)
        at org.flywaydb.core.internal.jdbc.JdbcTemplate.executeStatement(JdbcTemplate.java:201)
        at org.flywaydb.core.internal.sqlscript.ParsedSqlStatement.execute(ParsedSqlStatement.java:95)
        at org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor.executeStatement(DefaultSqlScriptExecutor.java:210)
        ... 159 common frames omitted
```

### CODE REVIEW

**Comment by tschaffter** [comment]
## Path param as a String OR Integer

The REST API server generated does not support specifying an integer when the parameter is defined as a string in the API description.

Example:

```console
GET {{basePath}}/organizations/1

HTTP/1.1 500 
{
  "title": "An exception occured",
  "status": 500,
  "detail": null,
  "type": null
}
```

The same occurs when using Swagger UI:

image

**Comment by tschaffter** [comment]
## anyOf

I tried to define the path parameter as:

```yaml
name: org
in: path
description: The unique identifier of the organization.
required: true
schema:
  oneOf:
    - $ref: ../../schemas/OrganizationLogin.yaml
    - $ref: ../../schemas/OrganizationId.yaml
```

The OA generator for Java replace the initial string type by creating this interface.

> **Note** The name of this interface is not very elegant. Another issue is that this interface is used in code generated in the `api` folder but this interface is not imported by default.

```java
package org.sagebionetworks.openchallenges.organization.service.model.dto;

import java.util.*;
import javax.annotation.Generated;
import javax.validation.constraints.*;

@Generated(value = "org.openapitools.codegen.languages.SpringCodegen")
public interface GetOrganizationOrgParameterDto {}
```

File: libs/openchallenges/api-description/src/components/parameters/path/org.yaml [hunk 1]
```diff
@@ -1,6 +1,13 @@
 name: org
 in: path
-description: The unique identifier of the organization.
+description: The id or login of the organization.
 required: true
 schema:
-  $ref: ../../schemas/OrganizationLogin.yaml
+  type: string
+  minLength: 1
+  maxLength: 25
```
Comment (rrchai) [inline libs/openchallenges/api-description/src/components/parameters/path/org.yaml:8]: some names of orgs in current database are longer than 25 characters (i.e `nsia-kano-diagnostic-center`)- suggest using longer maxLength.

**Review by rrchai — CHANGES_REQUESTED** [review]

File: libs/openchallenges/api-description/src/components/parameters/path/org.yaml (hunk 1 above)
Comment (tschaffter) [reply libs/openchallenges/api-description/src/components/parameters/path/org.yaml:8]: Good catch! They should be 2-64 chars by the decision documented in this [ticket](https://github.com/Sage-Bionetworks/sage-monorepo/pull/1482).

Comment (tschaffter) [reply libs/openchallenges/api-description/src/components/parameters/path/org.yaml:8]: Here we need to support 1 as minlength to support org ID 0-9.

Comment (tschaffter) [reply libs/openchallenges/api-description/src/components/parameters/path/org.yaml]: This issue is fixed

**Review by rrchai — APPROVED** [review]
LGTM! Thanks.

**Pull request merged**